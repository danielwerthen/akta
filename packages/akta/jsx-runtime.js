export { Fragment, jsx, jsxs } from './dist';
