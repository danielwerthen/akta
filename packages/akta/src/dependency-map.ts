import { BehaviorSubject, Observable, of } from 'rxjs';
import { switchMap } from 'rxjs/operators';

type EagerDependency<T> = {
  key: symbol;
  value: T;
};
type LazyDependency<T> = {
  key: symbol;
  init: () => T;
};

export type Dependency<T> = EagerDependency<T> | LazyDependency<T>;

export function isLazyDependency<T>(
  dep: Dependency<T>
): dep is LazyDependency<T> {
  return typeof (dep as LazyDependency<T>).init === 'function';
}

const LocalStop = Symbol('Dependency stopper');

export function getInitialValue<T>(dep: Dependency<T>): T {
  if (isLazyDependency(dep)) {
    return dep.init();
  }
  return dep.value;
}

export function createDependency<T>(initialValue: T): Dependency<T> {
  return {
    key: Symbol('Dependency'),
    value: initialValue,
  };
}

export function createLazyDependency<T>(init: () => T): Dependency<T> {
  return {
    key: Symbol('Dependency'),
    init,
  };
}

export class DependencyMap {
  private store: Map<Symbol, BehaviorSubject<unknown>> = new Map();
  parent?: DependencyMap;
  constructor(parent?: DependencyMap) {
    this.parent = parent;
  }

  private getLocal<T>(dep: Dependency<T>): BehaviorSubject<T> {
    const { key } = dep;
    const subject = this.store.get(key) ?? new BehaviorSubject(LocalStop);
    if (!this.store.has(key)) {
      this.store.set(key, subject);
    }
    return subject as BehaviorSubject<T>;
  }
  observe<T>(dep: Dependency<T>): Observable<T> {
    const subject = this.getLocal(dep);
    return subject.pipe(
      switchMap(val => {
        if ((val as unknown) === LocalStop) {
          return this.parent
            ? this.parent.observe(dep)
            : of(getInitialValue(dep));
        }
        return of(val);
      })
    );
  }
  peek<T>(dep: Dependency<T>): T {
    const value = this.getLocal(dep).value as unknown;
    if (value !== LocalStop) {
      return value as T;
    }
    if (this.parent) {
      return this.parent.peek(dep);
    }
    return getInitialValue(dep);
  }
  branch(): DependencyMap {
    return new DependencyMap(this);
  }
  provide<T>(dep: Dependency<T>, value: T | ((old?: T) => T)) {
    const subject = this.getLocal(dep);
    if (typeof value === 'function') {
      const fn = value as <T>(old?: T) => T;
      subject.next(fn(this.peek(dep) ?? undefined));
    } else {
      subject.next(value);
    }
  }
  unprovide(dep: Dependency<unknown>): unknown {
    const subject = this.getLocal(dep);
    const prev = subject.value;
    if (prev !== LocalStop) {
      subject.next(LocalStop);
      return prev;
    }
    return;
  }
}
