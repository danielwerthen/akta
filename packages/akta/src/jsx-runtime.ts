// Based on https://raw.githubusercontent.com/DefinitelyTyped/DefinitelyTyped/314d0c3cf25e1762525a3790b4609eb19fddadf9/types/react/index.d.ts

import { AktaElement, AktaElementType } from './types';

export function jsx<
  P extends {},
  T extends string | ((props: P) => AktaElement<P, any>)
>(type: T, props: P, maybeKey?: string): AktaElement<P, T> {
  return {
    $$type: AktaElementType,
    type,
    props,
    key: maybeKey,
  };
}

export const jsxs = jsx;
