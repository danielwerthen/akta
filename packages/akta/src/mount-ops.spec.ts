import { from, of, Subject } from 'rxjs';
import { DependencyMap } from './dependency-map';
import { jsx } from './jsx-runtime';
import {
  getPrev,
  LazyAttacher,
  observeNode,
  mount,
  NodeObserver,
} from './mount-ops';

describe('Mount ops 2', () => {
  let attacher = new LazyAttacher();
  const root = document.createElement('article');
  beforeEach(() => {
    root.innerHTML = '';
    attacher = new LazyAttacher();
  });

  it('should create deterministic structures', () => {
    attacher.attach(document.createElement('div'), [0]);
    attacher.attach(document.createTextNode('Foobar'), [1, 1]);
    attacher.attach(document.createTextNode('Baz'), [1, 0]);
    attacher.activate(node => root.append(node));
    attacher.attach(document.createElement('div'), [2]);
    expect(root).toMatchSnapshot();
  });
});

describe('Observe node', () => {
  it('shoudl work', () => {
    const deps = new DependencyMap();
    const attacher = new LazyAttacher();
    const root = document.createElement('article');
    function Component({ children }: any) {
      return children;
    }
    const observer = new NodeObserver();
    observeNode(
      of(
        jsx('div', {
          children: ['baz', jsx(Component, { children: 'test' }), of('foobar')],
        })
      ),
      deps,
      attacher,
      observer
    );
    observer.observe().subscribe(() => {
      attacher.activate(node => root.append(node));
    });
    expect(root).toMatchSnapshot();
  });

  it('should place later nodes in right order', () => {
    const root = document.createElement('article');
    mount(
      jsx('div', {
        children: [
          jsx('div', {
            children: [from(['hide', 'first']), 'second', 'third'],
          }),
          jsx('div', {
            children: ['first', from(['hide', 'second']), 'third'],
          }),
          jsx('div', {
            children: ['first', 'second', from(['hide', 'third'])],
          }),
        ],
      }),
      root
    );
    expect(root).toMatchSnapshot();
  });

  it('should handle case 4', () => {
    const root = document.createElement('article');
    function Daniel() {
      throw new Error('Foob');
      return 'comp';
    }
    try {
      mount(
        jsx('div', {
          children: [
            jsx('div', {
              children: [
                jsx('div', {
                  children: [
                    jsx('div', {
                      children: [jsx(Daniel, {})],
                    }),
                  ],
                }),
              ],
            }),
          ],
        }),
        root
      );
      throw new Error('Should not end up here');
    } catch (e) {
      const error = e as Error;
      const stackDepth = ((error.stack || '').match(/at observeNode/g) || [])
        .length;
      expect(stackDepth).toEqual(1);
      expect(root).toMatchSnapshot();
    }
  });

  it('should handle fragments properly', () => {
    const root = document.createElement('article');
    function Comp() {
      return jsx('div', {
        children: ['test', 'foo'],
      });
    }
    function App() {
      return jsx('div', {
        children: [
          jsx('h2', {
            children: 'Header',
          }),
          jsx(undefined as any, {
            children: [jsx(Comp, {}), jsx(Comp, {})],
          }),
          jsx('div', {
            children: 'after',
          }),
        ],
      });
    }
    mount(jsx(App, {}), root);
    expect(root).toMatchSnapshot();
  });

  it('should mount elements in correct order', () => {
    const root = document.createElement('article');
    const momo = new Subject();
    function App() {
      return jsx('div', {
        children: [
          of('Param id: 42'),
          momo,
          of(null),
          from([null, 'Match 42']),
        ],
      });
    }
    mount(jsx(App, {}), root);
    momo.next('Foobar2');
    expect(root).toMatchSnapshot();
  });
});

describe('getPrev', () => {
  it('should generate indicies in the right order', () => {
    const foo = getPrev([2, 3]);
    expect(foo.next().value).toStrictEqual([2, 2]);
    expect(foo.next().value).toStrictEqual([2, 1]);
    expect(foo.next().value).toStrictEqual([2, 0]);
    expect(foo.next().value).toStrictEqual([1]);
    expect(foo.next().value).toStrictEqual([0]);
    expect(foo.next().done).toBe(true);
  });
});
