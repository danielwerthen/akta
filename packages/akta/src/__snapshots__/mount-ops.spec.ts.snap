// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mount ops 2 should create deterministic structures 1`] = `
<article>
  <div />
  Baz
  Foobar
  <div />
</article>
`;

exports[`Observe node shoudl work 1`] = `
<article>
  <div>
    baz
    test
    foobar
  </div>
</article>
`;

exports[`Observe node should handle case 4 1`] = `<article />`;

exports[`Observe node should handle fragments properly 1`] = `
<article>
  <div>
    <h2>
      Header
    </h2>
    <div>
      test
      foo
    </div>
    <div>
      test
      foo
    </div>
    <div>
      after
    </div>
  </div>
</article>
`;

exports[`Observe node should place later nodes in right order 1`] = `
<article>
  <div>
    <div>
      first
      second
      third
    </div>
    <div>
      first
      second
      third
    </div>
    <div>
      first
      second
      third
    </div>
  </div>
</article>
`;
